plugins/kafka/kafka.cpp

```diff
@@ -0,0 +1,1118 @@
+                        lockPtr->unlock();
+                        throw;
+                    }
+                }
+
+                lockPtr->unlock();
+
+                if (!pubObjPtr)
+                {
+                    throw MakeStringException(-1, "Kafka: Unable to create publisher for brokers '%s' and topic '%s'", _brokers.c_str(), _topic.c_str());
+                }
+
+                return pubObjPtr;
+            }
+
+        private:
+
+            ObjMap      cachedPublishers;   //!< std::map of created Publisher object pointers
+            Mutex*      lockPtr;            //!< Mutex guarding modifications to cachedPublishers
+            __int32     traceLevel;         //!< The current logging level
+    } *publisherCache;
+
+    //--------------------------------------------------------------------------
+
+    /** @class  PublisherCacheExpirerObj
+     *          Class used to expire old publisher objects held within publisherCache
+     */
+    static class PublisherCacheExpirerObj : public Thread
+    {
+        public:
+
+            PublisherCacheExpirerObj()
+                :   Thread("Kafka::PublisherExpirer")
+            {
+                shouldRun = false;
+            }
+
+            virtual ~PublisherCacheExpirerObj()
+            {
+
+            }
+
+            virtual void start()
+            {
+                if (!isAlive())
+                {
+                    shouldRun = true;
+                    Thread::start();
+                }
+            }
+
+            virtual void stop()
+            {
+                if (isAlive())
+                {
+                    shouldRun = false;
+                    join();
+                }
+            }
+
+            virtual int run()
+            {
+                while (shouldRun)
+                {
+                    if (publisherCache)
+                    {
+                        publisherCache->expire();
+                    }
+
+                    usleep(1000);
+                }
+
+                return 0;
+            }
+
+        private:
+
+            bool        shouldRun;      //!< If true, we should execute our thread's main event loop
+    } *publisherCacheExpirer;
+
+    //--------------------------------------------------------------------------
+    // Lazy Initialization
+    //--------------------------------------------------------------------------
+
+    /**
+     * Make sure the publisher object cache is initialized as well as the
+     * associated background thread for expiring idle publishers.  Note that
+     * we're using one conditional (the presence of the cache object) as
+     * the test for both the cache object and the background thread.
+     */
+    static void ensurePublisherCacheSetup(__int32 _traceLevel)
+    {
+        if (!KafkaPlugin::publisherCache)
+        {
+            CriticalBlock   block(initLock);
+
+            if (!KafkaPlugin::publisherCache)
```

Comment:
This implementation of double checking locking is actuall thread safe.  The C++ compiler can reuse the value  that was calculated before the critical block inside the critical block.
(E.g., google search for double checked locking is broken).  
It can be fixed by using std::atomic for publisherCache (or possibly volatile, but std::atomic is cleaner)
.